### PR TITLE
Basic Image Creation

### PR DESCRIPTION
**Pull request by JBorrow**
This is my attempt at implementing a basic SPH projection (scatter) using `numba`.

The new functions in `visualisation.py` allow you to make projections of a whole `SWIFTDataset`. Nothing more, nothing less -- this is not intended to be a full replacement for any rendering libraries out there. Just a little experiment.

### CODE REVIEW

**Comment by JBorrow** [comment]
Today is a good day
```
Generating particles
Finished generating particles
Compiling
Took 1.1308925151824951 to compile.
Scattering
Took 16.765869617462158 to scatter.
Comparing with pySPHViewer
pySPHViewer took 218.46879053115845 on the same problem.
Note that pySPHViewer is running in single-threaded mode.
That makes us 13.030567188928668 x faster 😀 
```

File: scripts/benchmark_visualisation.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    h = 1.778_002 * h  # The kernel_gamma we use.
+
+    t = time()
+    qv = QuickView(
+        coordinates,
+        hsml=h,
+        mass=m,
+        x_size=res,
+        y_size=res,
+        r="infinity",
```
Comment (AshKelly) [inline scripts/benchmark_visualisation.py]: I can't quite remember but I think these should  be `xsize` and `ysize` and the default to 800. In that case, your code might be even faster than you think in comparison

File: swiftsimio/visualisation.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+@jit(nopython=True)
+def pair_min(a: int32, b: int32) -> int32:
+    if a < b:
+        return a
+    else:
+        return b
+
+
+@jit(nopython=True, fastmath=True)
+def scatter(x: float32, y: float32, m: float32, h: float32, res: int) -> ndarray:
```
Comment (AshKelly) [inline swiftsimio/visualisation.py]: I presume the rest of swiftsimio is float32?

File: scripts/benchmark_visualisation.py (hunk 1 above)
Comment (JBorrow) [reply scripts/benchmark_visualisation.py]: You’re right, thanks! The dangers of **kwargs, I guess

File: swiftsimio/visualisation.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    # Change that integer to a float, we know that our x, y are bounded
+    # by [0, 1].
+    float_res = float32(res)
+    pixel_width = 1.0 / float_res
+
+    # If the kernel width is smaller than this, we drop to just PIC method
+    drop_to_single_cell = pixel_width * 0.5
+
+    # Pre-calculate this constant for use with the above
+    inverse_cell_area = res * res
```
Comment (AshKelly) [inline swiftsimio/visualisation.py]: this is the number of pixels?

File: swiftsimio/visualisation.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            # The number of cells that this kernel spans
+            cells_spanned = int32(1.0 + kernel_width * float_res)
+
+            # Now we loop over the square of cells that the kernel lives in
+            for cell_x in range(
+                # Ensure that the lowest x value is 0, otherwise we'll segfault
+                pair_max(0, particle_cell_x - cells_spanned),
+                # Ensure that the highest x value lies within the array bounds,
+                # otherwise we'll segfault (oops).
+                min(particle_cell_x + cells_spanned, maximal_array_index),
```
Comment (AshKelly) [inline swiftsimio/visualisation.py]: why not use pair min since you defined it?

File: swiftsimio/visualisation.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            ):
+                # The distance in x to our new favourite cell -- remember that our x, y
+                # are all in a box of [0, 1].
+                distance_x = float32(cell_x) * pixel_width - x_pos
+                distance_x_2 = distance_x * distance_x
+                for cell_y in range(
+                    pair_max(0, particle_cell_y - cells_spanned),
+                    min(particle_cell_y + cells_spanned, maximal_array_index),
```
Comment (AshKelly) [inline swiftsimio/visualisation.py]: again, `pair_min`?

File: scripts/benchmark_visualisation.py (hunk 1 above)
Comment (AshKelly) [reply scripts/benchmark_visualisation.py]: Yea, I've used a library that did a clever thing where they compared the **kwargs to those expected using inspection like `__code__.co_varnames` and raise an error if you passed in something unexpected

**Comment by AshKelly** [comment]
I presume the images generated by pysphviewer and your in-house code look similar?

File: swiftsimio/visualisation.py (hunk 2 above)
Comment (JBorrow) [reply swiftsimio/visualisation.py]: That's a good point. This should actually be `float64` as we store positions as doubles.

File: swiftsimio/visualisation.py (hunk 4 above)
Comment (JBorrow) [reply swiftsimio/visualisation.py]: That's an excellent point. I have no idea how this worked before.

**Comment by AshKelly** [comment]
btw, this looks good now!

**Comment by JBorrow** [comment]
Thanks ash! At the moment I'm only supporting square images of the whole box, hence why there is only one DPI parameter.

Cheers again for the review!

**Pull request merged**